pyannote/audio/utils/metric.py

```diff
@@ -177,13 +192,29 @@ def der_from_swf(
 
         # if (num_frames, num_speakers)-shaped, compute just one DER for the whole file
         if ndim == 2:
-            return self.compute_components_helper(hypothesis.data, reference.data)
+            if uem is not None:
+                components = self.init_components()
+                for segment in uem:
+                    h = hypothesis.crop(segment)
+                    if h.size == 0:
+                        continue
```

Comment:
Why this corner case?

Reply:
For example:
```python
uem = Timeline([Segment(0, 3), Segment(10, 12)])
hypothesis.extent == reference.extent == Segment(0, 5)
```
If we crop `hypothesis` and `reference` on `uem[1]` we get an empty array, so calculating its components will cause a crash.
Now that I think about it, a more efficient option would be:
```suggestion
                    if hypothesis.extent.start >= window.end:
                        break
                    elif hypothesis.extent.end <= window.start:
                        continue
```
before cropping `hypothesis`.

What do you think?

Reply:
Or check that `hypothesis.extent` covers `uem` before the loop and raise a ValueError otherwise.

Reply:
Not really, they could very well not fully cover each other and we could still find a common region:
```python
uem = Timeline([Segment(0, 3)])
hypothesis.extent = Segment(1, 4)
```
In this case, `Timeline.covers()` doesn't work either way but we can still compute the DER over `Segment(1, 3)`.

The "break/continue" approach avoids having to loop multiple times over the uem, but I agree that we should also check at the end if no components have been calculated (i.e. no overlap between `hypothesis.extent` and `uem`) and raise a ValueError if needed.

Reply:
In `pyannote.metrics` philosophy, 

* when a `uem` is provided, it is expected that its whole support has been processed by the evaluated pipeline (and hence should be available in `hypothesis`)
* when none is provided, [it is inferred automatically from the union of `reference` and `hypothesis` extents](https://github.com/pyannote/pyannote-metrics/blob/a4fa3d351dc9e3fc5aba216b89853c51573695e1/pyannote/metrics/utils.py#L175-L185)

Raise a `ValueError` when `hypothesis.extent` does not cover `uem` seems more consistent with the behavior of other metrics.

Reply:
Yes.

But you should use `hypothesis.sliding_window[0].start` --> `hypothesis.sliding_window[num_chunks - 1].end` instead of `hypothesis.extent` because of the weird implementation of `extent` that actually returns `hypothesis.sliding_window[0].middle` -> `hypothesis.sliding_window[num_chunks - 1].middle`

Reply:
but we're talking about 2D sliding window features here, so isn't `num_chunks == 1`?

Reply:
Oh, right. `sed s/num_chunks/num_frames`.

Reply:
I've added the code, let me know if you want to do any other changes